### ISSUE TITLE
Sensors with variable data (like analog 0-5v and CAN data) should have some sort of smoothing

### ISSUE THREAD

**Issue by ElDominio** [rusefi/rusefi#6579]
**Describe the bug**
Some senors, like in this case AFR,  have high variability. In this case, a V-Twin has twin wideband O2 sensors, but the widebands show the lean spots in between cylinders firing. This should be smoothed out by software as to not throw off anything that depends on these sensors for PID stuff like closed loop AFR (I am thinking this makes P-term quite unstable)

**Hardware**
 - Board: Harley PNP
 - Engine info Harely vtwin

**Expected behavior**
I have a filter to smooth out the data for Air-Fuel ratio shown on screenshot, making it read a more stable value

**Screenshots**
![image](https://github.com/rusefi/rusefi/assets/5685108/8f196a95-18a6-48db-9864-5acb13fcedc2)

**Comment by mck1117** [comment]
Those are not lean spots between firings. There's still exhaust in the pipe. The sensor also doesn't have anywhere near enough bandwidth to resolve individual cylinders at 6000 rpm.

Do you have any data to support that this variation isn't real?

**Comment by mck1117** [comment]
I bet that variation is real, just look at your MAP!

**Comment by ElDominio** [comment]
> Those are not lean spots between firings. There's still exhaust in the pipe. The sensor also doesn't have anywhere near enough bandwidth to resolve individual cylinders at 6000 rpm.
> 
> Do you have any data to support that this variation isn't real?

I'll grab the log from their stuff

This is a V Twin, with individual exhausts per cylinder (i did mention it being a vtwin), and MAP reads from one cylinder (i increased sample time to 150ms at 540 degrees). I think it's not totally unreasonable to think we can see one cylinder reflected on the sheet? But idk

**Comment by mck1117** [comment]
> I think it's not totally unreasonable to think we can see one cylinder reflected on the sheet

But what other gas would there be in the pipe? How would there be something other than exhaust from the previous firing near the sensor?

**Comment by mck1117** [comment]
> I think it's not totally unreasonable to think we can see one cylinder reflected on the sheet

6000 rpm = 100hz = 1 cylinder firing every 10ms

**Comment by ElDominio** [comment]
assuming an pressure wave can travel at the speed of sound; 1,125ft/s (let's say slower for te sake of it, 1000ft/s), so 1ft/ms, on a 3 foot long pipe (this is a bike, where the exhaust is halfway of the chassis, probably about a foot long but 3ft to give it favorable numbers) the exhaust gas would leave the exhaust with about 15ms to spare (3foot pipe * 1ft/ms = 3ft of travel in 3ms), which I also think is ample  time for fresh air to come into the pipe and start finding its way into the sensing element of the 4.9

**Comment by ElDominio** [comment]
Log had min/max values for AFR at 10/16, meaning small variations look enormous. Here is the log
[2024-06-03_14.41.44wot7.mlg.zip](https://github.com/user-attachments/files/15574507/2024-06-03_14.41.44wot7.mlg.zip)

I don't think it's that critical but idk

**Comment by benassi124** [comment]
> Log had min/max values for AFR at 10/16, meaning small variations look enormous. Here is the log [2024-06-03_14.41.44wot7.mlg.zip](https://github.com/user-attachments/files/15574507/2024-06-03_14.41.44wot7.mlg.zip)
> 
> I don't think it's that critical but idk

3 out of the 7 units that have gone out so far, the customers have complained about this, they arent used to this, its been a hard sell on that aspect so far

**Comment by ElDominio** [comment]
Being not used to it doesn't make it incorrect. What bad effects does the AFR not being a smooth line cause the bike? Powerband looks fine, response looks fine. 3 out of the 7 people working on them are used to old, slow unresponsive analog widebands that update 10 times per second at best

**Comment by benassi124** [comment]
@ElDominio not sure if this helps at all, two exactly the same engines, the left chart is maxxecu, which is currently considered the gold standard, by hardcore harley/bike racers, on the right is our ECU.
The left is what most of these guys are expecting, I am nowhere near smart enough to know why the left with maxxecu is so much smoother and less jumpy, however, I am hoping to learn
![image](https://github.com/rusefi/rusefi/assets/146249887/06ed6ac7-a608-408e-8d98-ef88db8e1395)

**Comment by ElDominio** [comment]
Can you provide the log itself from the MaxxECU?

**Comment by benassi124** [comment]
[Suburban HD Bridgestone Bike_10539last_10545.zip](https://github.com/user-attachments/files/15596080/Suburban.HD.Bridgestone.Bike_10539last_10545.zip)

**Comment by benassi124** [comment]
> [Suburban HD Bridgestone Bike_10539last_10545.zip](https://github.com/user-attachments/files/15596080/Suburban.HD.Bridgestone.Bike_10539last_10545.zip)

@ElDominio

**Comment by benassi124** [comment]
1 CAN WBO2, 1 Analog. Lambda A on the left is analog, Lambda B on the right is CAN.

Im not as much worried about the lambda as I am to lambda reacting to those crazy fluctuations in MAP, I was trying to illustrate that its not an engine thing, but something else potentially.
Or maybe im dumb, which is also very likely but willing to learn lol

![image](https://github.com/rusefi/rusefi/assets/146249887/14185698-fdcb-4ba3-9c24-2f12a2030eb2)

**Comment by benassi124** [comment]
here is a log from a maxxecu without that odd-fire map filter
![image](https://github.com/rusefi/rusefi/assets/146249887/7af6216c-3f92-418c-92d2-694e4386baf6)

**Comment by ElDominio** [comment]
I _think_ this is the purpose of MAP sampling angle and window; but we'd have to play around with it to fix the MAP reading so it looks less jumpy

**Comment by benassi124** [comment]
> I _think_ this is the purpose of MAP sampling angle and window; but we'd have to play around with it to fix the MAP reading so it looks less jumpy



> I _think_ this is the purpose of MAP sampling angle and window; but we'd have to play around with it to fix the MAP reading so it looks less jumpy

Correct I agree, we could possibly code in different levels of map sampling so it can be user friendly or a "User Defined Setting" where you can manually set them for more advanced users such as yourself.

Because when uncle fester and his 3 hillbilly mechanics in the hills of west virgina get their hands on it theyre not going to sit there and dial it in lol. It will help overall UX of RusEFI

**Comment by benassi124** [comment]
@ElDominio 
Adding basic smoothing factor of 5 made this log much more readable by a tuner imo, 

![image](https://github.com/rusefi/rusefi/assets/146249887/6a687947-5726-4146-a1b0-751b4e16663c)

**Comment by rusefillc** [comment]
@benassi124 sounds like you want the level of smoothness similar to https://rusefi.com/online/view.php?log=1405 ?

**Comment by rusefillc** [comment]
@ElDominio @benassi124 shall we try different numbers up to 12?

![image](https://github.com/rusefi/rusefi/assets/48498823/47a3c043-4c7f-479f-8856-8dd691c42f08)

**Comment by rusefillc** [comment]
@benassi124 @ElDominio let's start with MAP exponential average

![image](https://github.com/rusefi/rusefi/assets/48498823/96b34b93-6fc2-4143-8bb7-32d66ac65890)

left has default value '1' while right has alpha of 0.4

![image](https://github.com/rusefi/rusefi/assets/48498823/22a07ce4-6f43-4056-83bc-db5a233963cb)

**Comment by benassi124** [comment]
Thank you

**Comment by zed65** [comment]
If you set the MAP exponential average Alpha to a bigger value than 1 it throws a CRC check error. Maybe a MAX value of 1 should be implemented?

**Comment by rusefillc** [comment]
@zed65 yes https://github.com/rusefi/rusefi/commit/f7d5d48d00fe7b507ed88da66236193a295ffb89
